### PR TITLE
feat: add patch-based GitLab issue description updates (#471)

### LINKED ISSUE #471: Add patch-based GitLab issue description updates
## Summary

I would like to request support for updating GitLab issues using a diff/patch-based workflow instead of requiring the MCP client to send the full issue description every time.

Currently, when an AI agent wants to update a long issue description, it appears that it needs to provide the entire updated issue body. This can consume a large number of tokens, especially for issues with long markdown descriptions, checklists, logs, acceptance criteria, or generated content.

A more token-efficient workflow would be to let the MCP server fetch the current issue description, apply a small patch/diff locally, and then submit the final updated description to the GitLab API.

## Problem

When updating a GitLab issue, the agent often only needs to make a small change, for example:

- add one checkbox item
- update a status section
- append a note
- replace a paragraph
- fix a typo
- update acceptance criteria
- add a progress comment into the issue description

However, without a patch-based tool, the agent must send the entire issue description back through the MCP call.

This has several downsides:

1. High token usage for long issues.
2. Greater risk of accidentally dropping existing content.
3. Worse UX for agents working with large markdown issue descriptions.
4. More difficult review/debugging because the intended change is not isolated.

## Proposed solution

Add a new MCP tool for patch-based issue description updates.

Possible tool name:

```text
update_issue_description_patch
````

or:

```text
update_issue_with_patch
```

### Suggested input schema

```json
{
  "project_id": "string | number",
  "issue_iid": "number",
  "patch": "string",
  "patch_type": "unified_diff | search_replace",
  "dry_run": "boolean",
  "create_note": "boolean"
}
```

### Example: unified diff

```json
{
  "project_id": "12345678",
  "issue_iid": 42,
  "patch_type": "unified_diff",
  "dry_run": false,
  "patch": "--- old\n+++ new\n@@\n-Status: In progress\n+Status: Ready for review\n"
}
```

### Example: search/replace block

```json
{
  "project_id": "12345678",
  "issue_iid": 42,
  "patch_type": "search_replace",
  "dry_run": false,
  "patch": "<<<<<<< SEARCH\nStatus: In progress\n=======\nStatus: Ready for review\n>>>>>>> REPLACE"
}
```

## Expected behavior

The MCP server should:

1. Fetch the current issue using the GitLab API.
2. Read the current issue description.
3. Apply the provided patch locally.
4. Validate that the patch applies cleanly.
5. Submit the full resulting description to GitLab using the existing issue update API.
6. Return a concise result showing what changed.

## Error handling

The tool should fail safely if:

* the issue does not exist
* the issue has no description and the patch requires matching existing text
* the patch cannot be applied
* the search block is ambiguous
* the search block does not match
* the resulting description is unchanged
* GitLab rejects the update

For `search_replace`, I think it should fail if the search text matches more than once, unless an explicit option is provided to allow multiple replacements.

## Dry-run support

It would be useful to support `dry_run: true`.

In dry-run mode, the tool would:

* fetch the current issue
* apply the patch locally
* return whether the patch applies successfully
* return a short preview/diff of the resulting change
* not update the GitLab issue

This would let the agent verify the change before applying it.

## Optional: note/comment mode

An optional `create_note` flag could make the server add a comment/note to the issue summarizing the patch that was applied.

For example:

```text
Updated issue description using patch-based MCP tool.
Changed:
- Status: In progress -> Ready for review
```

This is not required, but it could be useful for auditability.

## Why this should live in the MCP server

GitLab's issue update API still requires sending the final full description, but the MCP client/agent should not need to provide the full description.

The MCP server can act as the safe intermediary:

```text
Agent sends small patch
        ↓
MCP server fetches current issue
        ↓
MCP server applies patch
        ↓
MCP server sends final description to GitLab
```

This keeps the token usage low while still using the existing GitLab API correctly.

## Use case

I use this MCP server with AI coding agents. Some GitLab issues contain long markdown descriptions. When the agent only needs to change a small section, sending the entire issue description is inefficient.

A patch-based update tool would make issue editing much more token-efficient for long-running agent workflows.

### PR DESCRIPTION
**Pull request by zereight**
## Summary

Adds `update_issue_description_patch` MCP tool that applies search/replace or unified diff patches to issue descriptions. This reduces token usage for AI agents by allowing small changes without sending the full description.

## Changes

- **New schema**: `UpdateIssueDescriptionPatchSchema` in `schemas.ts`
- **Patch helper**: `utils/patch-helper.ts` — parseSearchReplaceBlocks, applySearchReplace, applyUnifiedDiff
- **Tool handler**: `index.ts` — getIssue → patch → dry_run or PUT → optional create_note
- **Registry**: `tools/registry.ts` — tool definition + issues toolset
- **Mock server**: `test/utils/mock-gitlab-server.ts` — PUT issues endpoint + in-memory store
- **Tests**: 12 unit tests covering all patch formats and edge cases
- **Docs**: `skills/gitlab-mcp/reference/issues.md` + `README.md` tool list

### search_replace (exact match)
```json
{
  "patch_type": "search_replace",
  "patch": "<<<<<<< SEARCH\nStatus: In progress\n=======\nStatus: Ready for review\n>>>>>>> REPLACE"
}
```

### unified_diff (standard diff format)
```json
{
  "patch_type": "unified_diff",
  "patch": "--- old\n+++ new\n@@ -1 +1 @@\n-Status: In progress\n+Status: Done"
}
```

### Options
- `dry_run: true` — preview changes, no PUT call
- `create_note: true` — auto-add change summary note
- `allow_multiple: true` — replace all occurrences (search_replace only)

### Error handling
- Search text not found → clear error
- Multiple matches without allow_multiple → error with count
- Identical result → no-op error
- Unified diff mismatch → error

## Verification
- ✅ TypeScript compiles clean
- ✅ `npm run build` succeeds
- ✅ 70 existing tests pass (no regression)
- ✅ 12 new unit tests pass
- Closes #471

### CODE REVIEW

File: utils/patch-helper.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+ *   =======
+ *   text to replace with
+ *   >>>>>>> REPLACE
+ *
+ * Supports multiple blocks.
+ */
+export function parseSearchReplaceBlocks(patch: string): SearchReplaceBlock[] {
+  const blocks: SearchReplaceBlock[] = [];
+  // Match SEARCH...REPLACE blocks (greedy multiline)
+  const regex = /<<<<<<< SEARCH\s*\n([\s\S]*?)=======\s*\n([\s\S]*?)>>>>>>> REPLACE\s*/g;
```
[the remark this answers is not in the document]
Comment (zereight) [reply utils/patch-helper.ts]: Resolved in be66999 (fix: change `\\s*` to `[^\\S\\n]*` to preserve leading blank lines)
